src/ReceiverLink.cs

```diff
@@ -97,29 +106,64 @@ public void Start(int credit, MessageCallback onMessage = null)
             this.SetCredit(credit, true);
         }
 
+        /// <summary>
+        /// Computes amount of credit to issue based on desired credit and
+        /// number of messages in flight. Will not return negative credit.
+        /// </summary>
+        /// <param name="count_rcv">reciever delivery count</param>
+        /// <param name="count_snd">sender deliver count</param>
+        /// <param name="credit">desirec message credit</param>
+        /// <returns></returns>
+        internal int ComputeCredit(int count_rcv, int count_snd, int credit)
+        {
+            uint rCredit = unchecked((uint)count_rcv + (uint)credit - (uint)count_snd);
+            int iCredit = Math.Max((int)rCredit, 0);
+            return iCredit;
+        }
+
+        /// <summary>
+        /// Compute receive delivery count at which autoRestore will
+        /// issue a new credit flow.
+        /// </summary>
+        /// <param name="count_rcv">receiver delivery count</param>
+        /// <param name="credit">credit setting</param>
+        /// <returns></returns>
+        internal int ComputeRestoreCount(int count_rcv, int credit)
+        {
+            int rCount = unchecked(count_rcv + ((credit + 1) / 2));
+            return rCount;
+        }
+
         /// <summary>
         /// Sets a credit on the link. A flow is sent to the peer to update link flow control state.
+        /// Lowering the credit value may result in send a flow with zero credits.
+        /// The autoRestore receive delivery count is computed.
         /// </summary>
         /// <param name="credit">The new link credit.</param>
-        /// <param name="autoRestore">If true, link credit is auto-restored when a message is accepted
-        /// or rejected by the caller. If false, caller is responsible for manage link credits.</param>
+        /// <param name="autoRestore"> If false, caller is responsible for
+        /// managing link credit. If true, link credit is auto-restored when
+        /// a message is disposed (Accept, Reject, Modified, Released) by the caller
+        /// and the receiver has disposed of enough messages so that a new credit of at least
+        /// credit/2 may be issued.</param>
         public void SetCredit(int credit, bool autoRestore = true)
         {
-            uint dc;
+            int flowCredit;
+            int dCountSnd;
             lock (this.ThisLock)
             {
                 if (this.IsDetaching)
                 {
                     return;
                 }
 
-                this.totalCredit = autoRestore ? credit : 0;
                 this.credit = credit;
-                this.restored = 0;
-                dc = this.deliveryCount;
+                this.autoRestore = autoRestore;
+                dCountSnd = deliveryCountSnd;
+                flowCredit  = ComputeCredit(deliveryCountRcv, deliveryCountSnd, credit);
```

Comment:
dCountSnd should be passed to ComputeCredit here rather than deliveryCountSnd, to ensure the values align precisely in the subsequent flow, as in DisposeMessage.